### PR TITLE
Improve the Converter interface

### PR DESCRIPTION
**Pull request by wixette**
* Supports Validate and Parse separately
* Parses inline text strings with invalid tokens
* Reports error info via Diagnostic collection
* Remove unused code in unittests

### CODE REVIEW

File: csharp/src/SeedLang/Block/Converter.cs [hunk 1]
```diff
@@ -23,44 +23,75 @@ namespace SeedLang.Block {
...
+
+      public void VisitInvalidToken(TextRange range) {
+        InvalidTokenRanges?.Add(range);
+      }
     }
 
-    // Converts an expression inline text to a list of blocks.
-    public static IEnumerable<BaseBlock> InlineTextToBlocks(string text) {
-      Debug.Assert(!string.IsNullOrEmpty(text), "Inline text shall not be null or empty.");
+    // Checks if an inline text is a valid expression. Detailed info will be stored in collection if
```
Comment (codingpotato) [inline csharp/src/SeedLang/Block/Converter.cs:65]: Detailed info -> Detailed diagnostic info.

File: csharp/src/SeedLang/Block/InlineTextParser.cs [hunk 2]
```diff
@@ -28,62 +29,72 @@ internal class InlineTextParser : BaseParser {
...
+            listener.VisitArithmeticOperator(token.Text, TextRange.RangeOfToken(token));
             break;
           case SeedBlockLexer.SUB:
             if (lastTokenType == SeedBlockParser.NUMBER ||
                 lastTokenType == SeedBlockParser.CLOSE_PAREN) {
-              listener.VisitArithmeticOperator(token.Text);
+              listener.VisitArithmeticOperator(token.Text, TextRange.RangeOfToken(token));
             } else {
               negative = true;
+              negativeToken = token;
```
Comment (codingpotato) [inline csharp/src/SeedLang/Block/InlineTextParser.cs:71]: negative and nagativeToken are duplicated. Is it better to remove negative variable and check if negativeToken is not null to see if met a negative sign?

File: csharp/src/SeedLang/Common/TextRange.cs [hunk 3]
```diff
@@ -50,5 +51,16 @@ public class TextRange : Range {
...
     }
+
+    public static TextRange RangeOfToken(IToken token) {
+      return RangeOfTokens(token, token);
+    }
+
+    public static TextRange RangeOfTokens(IToken start, IToken end) {
+      // TODO: need scan the source string to calculate the end column if the end token is in
+      // multiple lines.
+      return new TextRange(start.Line, start.Column,
```
Comment (codingpotato) [inline csharp/src/SeedLang/Common/TextRange.cs:62]: I feel it's not good that the TextRange class in Common component depends on Antlr4.Runtime. Maybe we can limit this kind of dependency in X and Block components. It's better to keep the Token to TextRange conversion in SeeLang.X or SeedLang.Block.

**Review by codingpotato — CHANGES_REQUESTED** [review]

File: csharp/src/SeedLang/Block/Converter.cs (hunk 1 above)
Comment (wixette) [reply csharp/src/SeedLang/Block/Converter.cs]: Done

File: csharp/src/SeedLang/Block/InlineTextParser.cs (hunk 2 above)
Comment (wixette) [reply csharp/src/SeedLang/Block/InlineTextParser.cs:69]: Done

File: csharp/src/SeedLang/Common/TextRange.cs (hunk 3 above)
Comment (wixette) [reply csharp/src/SeedLang/Common/TextRange.cs]: Done. Moved the two methods to a utility class in SeedLang.X.

**Review by codingpotato — APPROVED** [review]
LGTM

**Pull request merged**